pkg/apis/autoscaling/v2alpha1/defaults_test.go

```diff
@@ -0,0 +1,131 @@
+	"k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/api"
+	_ "k8s.io/kubernetes/pkg/api/install"
+	"k8s.io/kubernetes/pkg/apis/autoscaling"
+	_ "k8s.io/kubernetes/pkg/apis/autoscaling/install"
+	. "k8s.io/kubernetes/pkg/apis/autoscaling/v2alpha1"
+)
+
+func TestSetDefaultHPA(t *testing.T) {
+	utilizationDefaultVal := int32(autoscaling.DefaultCPUUtilization)
+	defaultReplicas := newInt32(1)
+	defaultTemplate := []autoscalingv2alpha1.MetricSpec{
+		{
+			Type: autoscalingv2alpha1.ResourceMetricSourceType,
+			Resource: &autoscalingv2alpha1.ResourceMetricSource{
+				Name: v1.ResourceCPU,
+				TargetAverageUtilization: &utilizationDefaultVal,
+			},
+		},
+	}
+
+	tests := []struct {
+		original *autoscalingv2alpha1.HorizontalPodAutoscaler
+		expected *autoscalingv2alpha1.HorizontalPodAutoscaler
+	}{
+		{ // MinReplicas default value
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					Metrics: defaultTemplate,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+		{ // MinReplicas update
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: newInt32(3),
+					Metrics:     defaultTemplate,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: newInt32(3),
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+		{ // Metrics default value
+			original: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+				},
+			},
+			expected: &autoscalingv2alpha1.HorizontalPodAutoscaler{
+				Spec: autoscalingv2alpha1.HorizontalPodAutoscalerSpec{
+					MinReplicas: defaultReplicas,
+					Metrics:     defaultTemplate,
+				},
+			},
+		},
+	}
+
+	for i, test := range tests {
+		original := test.original
+		expected := test.expected
+		obj2 := roundTrip(t, runtime.Object(original))
+		got, ok := obj2.(*autoscalingv2alpha1.HorizontalPodAutoscaler)
+		if !ok {
+			t.Fatalf("(%d) unexpected object: %v", i, got)
+		}
+		if !apiequality.Semantic.DeepEqual(got.Spec, expected.Spec) {
+			t.Errorf("(%d) got different than expected\ngot:\n\t%+v\nexpected:\n\t%+v", i, got.Spec, expected.Spec)
+		}
+	}
+}
+
+func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
+	data, err := runtime.Encode(api.Codecs.LegacyCodec(SchemeGroupVersion), obj)
+	if err != nil {
+		t.Errorf("%v\n %#v", err, obj)
+		return nil
+	}
+	obj2, err := runtime.Decode(api.Codecs.UniversalDecoder(), data)
+	if err != nil {
+		t.Errorf("%v\nData: %s\nSource: %#v", err, string(data), obj)
```

Comment:
data is []byte type here, here just print the data in string will more user-friendly, and the roundTrip function in every pkg/apis/GROUPS is same.